engine/timer.cpp

```diff
@@ -0,0 +1,50 @@
+#include "timer.h"
+
+#include <spdlog/spdlog.h>
+
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <memory>
+
+namespace pong
+{
+
+static constexpr int MS_TO_US = 1000;
+static constexpr int S_TO_US = 1000000;
+
+float Timer::frameTime = 1.0f;
+
+void Timer::HandleTimerCallbacks()
+{
+    const auto timeWaited = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now() - mLastTime);
+    mLastTime = std::chrono::system_clock::now();
+
+    Timer::frameTime = static_cast<float>(timeWaited.count()) / static_cast<float>(S_TO_US);
+
+    const int timeWaitedUs = static_cast<int>(timeWaited.count());
+    for (auto& timerRequest : mActiveTimers)
+    {
+        timerRequest->mTimeoutUs -= timeWaitedUs;
+        if (timerRequest->mTimeoutUs <= 0)
+        {
+            timerRequest->mCallback();
+        }
+    }
+
+    mActiveTimers.erase(std::remove_if(mActiveTimers.begin(), mActiveTimers.end(), [](const auto& timerRequest) {
+        return timerRequest->mTimeoutUs <= 0;
+    }), mActiveTimers.end());
+}
```

Comment:
Damn, look at all the modern cpp in this function. really a very different paradigm than spatial. I'm glad you're embracing it and learning it (regardless of if you like it or not). The experience with it is very valuable, and will be expected in many gigs.

Reply:
Yeah, so far it has been pretty fun to learn! Theres a lot of cool features in it. smart pointers are much, **much** nicer to use. Templates are sometimes annoying to deal with and figure out entirely, but they are very useful and I would much rather have them then not.